src/export/exported.c

```diff
@@ -35,6 +35,71 @@ sexp* rlang_interrupt() {
 }
 
 
+// dict.c
+
+sexp* rlang_new_dict(sexp* size, sexp* prevent_resize) {
+  if (!r_is_number(size)) {
+    r_abort("`size` must be an integer.");
+  }
+  if (!r_is_bool(prevent_resize)) {
+    r_abort("`prevent_resize` must be a logical value.");
+  }
+
+  sexp* out = KEEP(r_new_vector(r_type_list, 2));
+
+  struct r_dict dict = r_new_dict(r_int_get(size, 0));
```

Comment:
I think these PROTECT_ macros are a bit inefficient because you have to look them up each time, so I'm trying a new pattern where we store a protect handle in a `shelter` element of the struct. As long as you're protecting that element you're good.

By convention we could keep these as the first element of the struct. This has the somewhat interesting property that you can cast a pointer to the struct as a `sexp*` and store it directly in a list for protection (following C casting rules for structs).

Reply:
That sounds good, but don't you still need to `KEEP(dict.shelter)` here?

Reply:
oh absolutely! Done